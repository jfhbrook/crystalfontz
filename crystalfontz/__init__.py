# -*- coding: utf-8 -*-

"""
This is the documentation for crystalfontz.
"""

