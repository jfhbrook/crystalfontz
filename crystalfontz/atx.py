from dataclasses import dataclass
from enum import Enum
from typing import Iterable, Optional, Self

AUTO_POLARITY = 0x01


class AtxPowerSwitchFunction(Enum):
    RESET_INVERT = 0x02
    POWER_INVERT = 0x04
    LCD_OFF_IF_HOST_IS_OFF = 0x10
    KEYPAD_RESET = 0x20
    KEYPAD_POWER_ON = 0x40
    KEYPAD_POWER_OFF = 0x80


@dataclass
class AtxPowerSwitchFunctionalitySettings:
    functions: Iterable[AtxPowerSwitchFunction]
    auto_polarity: bool = True
    power_pulse_length_seconds: Optional[float] = None

    def to_bytes(self: Self) -> bytes:
        functions: int = 0
        for function in self.functions:
            functions ^= function.value
        if self.auto_polarity:
            functions ^= AUTO_POLARITY

        packed: bytes = functions.to_bytes()

        if self.power_pulse_length_seconds is not None:
            pulse_length = int(self.power_pulse_length_seconds * 32)

            if pulse_length < 1:
                raise ValueError(f"Pulse length can not be less than {1/32}")

            packed += min(pulse_length, 255).to_bytes()

        return packed
