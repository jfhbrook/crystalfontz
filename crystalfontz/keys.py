from dataclasses import asdict, dataclass
from enum import Enum
from typing import List, Literal, Self, Type

KeyPress = (
    Literal[0x01]
    | Literal[0x02]
    | Literal[0x04]
    | Literal[0x08]
    | Literal[0x10]
    | Literal[0x20]
)

KP_UP: KeyPress = 0x01
KP_ENTER: KeyPress = 0x02
KP_EXIT: KeyPress = 0x04
KP_LEFT: KeyPress = 0x08
KP_RIGHT: KeyPress = 0x10
KP_DOWN: KeyPress = 0x20


@dataclass
class KeyState:
    """
    A key's state.

    Attributes:
        pressed (bool): When True, the key is currently pressed.
        pressed_since (bool): When True, the key has been pressed since the last poll.
        released_since (bool): When True, the key has been released since the last
                               poll.
    """

    pressed: bool
    pressed_since: bool
    released_since: bool

    @classmethod
    def from_bytes(cls: Type[Self], state: bytes, keypress: KeyPress) -> Self:
        pressed = state[0]
        pressed_since = state[1]
        released_since = state[2]

        return cls(
            pressed=bool(pressed & keypress),
            pressed_since=bool(pressed_since & keypress),
            released_since=bool(released_since & keypress),
        )

    def to_bytes(self: Self, keypress: KeyPress) -> bytes: ...


@dataclass
class KeyStates:
    """
    The state of all keys.

    Attributes:
        up: The state of the "up" key.
        enter: The state of the "enter" key.
        exit: The state of the "exit" key.
        left: The state of the "left" key.
        right: The state of the "right" key.
        down: The state of the "down" key.
    """

    up: KeyState
    enter: KeyState
    exit: KeyState
    left: KeyState
    right: KeyState
    down: KeyState

    @classmethod
    def from_bytes(cls: Type[Self], state: bytes) -> Self:
        return cls(
            up=KeyState.from_bytes(state, KP_UP),
            enter=KeyState.from_bytes(state, KP_ENTER),
            exit=KeyState.from_bytes(state, KP_EXIT),
            left=KeyState.from_bytes(state, KP_LEFT),
            right=KeyState.from_bytes(state, KP_RIGHT),
            down=KeyState.from_bytes(state, KP_DOWN),
        )

    def to_bytes(self: Self) -> bytes:
        pressed = 0x00
        pressed_since = 0x00
        released_since = 0x00

        for state, keypress in [
            (self.up, KP_UP),
            (self.enter, KP_ENTER),
            (self.exit, KP_EXIT),
            (self.left, KP_LEFT),
            (self.right, KP_RIGHT),
            (self.down, KP_DOWN),
        ]:
            pressed = (pressed ^ keypress) if state.pressed else pressed
            pressed_since = (
                (pressed_since ^ keypress) if state.pressed_since else pressed_since
            )
            released_since = (
                (released_since ^ keypress) if state.released_since else released_since
            )

        return bytes([pressed, pressed_since, released_since])

    def __repr__(self: Self) -> str:
        repr_ = ""
        for name, state in asdict(self).items():
            st = ", ".join([f"{n}={'yes' if s else 'no'}" for n, s in state.items()])
            repr_ += f"{name}: {st}\n"
        return repr_[0:-1]


class KeyActivity(Enum):
    """
    A key activity. This is either a "press" event or a "release" event for a
    given key.
    """

    KEY_UP_PRESS = 1
    KEY_DOWN_PRESS = 2
    KEY_LEFT_PRESS = 3
    KEY_RIGHT_PRESS = 4
    KEY_ENTER_PRESS = 5
    KEY_EXIT_PRESS = 6
    KEY_UP_RELEASE = 7
    KEY_DOWN_RELEASE = 8
    KEY_LEFT_RELEASE = 9
    KEY_RIGHT_RELEASE = 10
    KEY_ENTER_RELEASE = 11
    KEY_EXIT_RELEASE = 12

    @classmethod
    def from_bytes(cls: Type[Self], activity: bytes) -> "KeyActivity":
        return KEY_ACTIVITIES[activity[0] - 1]


KEY_ACTIVITIES: List[KeyActivity] = [
    KeyActivity.KEY_UP_PRESS,
    KeyActivity.KEY_DOWN_PRESS,
    KeyActivity.KEY_LEFT_PRESS,
    KeyActivity.KEY_RIGHT_PRESS,
    KeyActivity.KEY_ENTER_PRESS,
    KeyActivity.KEY_EXIT_PRESS,
    KeyActivity.KEY_UP_RELEASE,
    KeyActivity.KEY_DOWN_RELEASE,
    KeyActivity.KEY_LEFT_RELEASE,
    KeyActivity.KEY_RIGHT_RELEASE,
    KeyActivity.KEY_ENTER_RELEASE,
    KeyActivity.KEY_EXIT_RELEASE,
]
