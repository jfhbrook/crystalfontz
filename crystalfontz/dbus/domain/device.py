from typing import ClassVar, Tuple

from crystalfontz.dbus.domain.base import (
    BytesM,
    BytesT,
    ModelM,
    ModelT,
    RevisionM,
    RevisionT,
    struct,
    t,
    Uint16M,
    Uint16T,
)
from crystalfontz.device import Device, DeviceStatus, lookup_device

WidthT = Uint16T


class WidthM(Uint16M):
    pass


HeightT = Uint16T


class HeightM(Uint16M):
    pass


DeviceT = Tuple[ModelT, RevisionT, RevisionT, WidthT, HeightT, WidthT, HeightT, Uint16T]


class DeviceM:
    t: ClassVar[str] = struct(
        ModelM, RevisionM, RevisionM, WidthM, HeightM, WidthM, HeightM, Uint16M
    )

    @staticmethod
    def pack(device: Device) -> DeviceT:
        return (
            device.model,
            device.hardware_rev,
            device.firmware_rev,
            device.lines,
            device.columns,
            device.character_width,
            device.character_height,
            device.n_temperature_sensors,
        )

    @staticmethod
    def unpack(device: DeviceT) -> Device:
        model = device[0]
        hardware_rev = device[1]
        firmware_rev = device[2]

        return lookup_device(model, hardware_rev, firmware_rev)


# Includes enough device info to unpack the status
DeviceStatusT = Tuple[ModelT, RevisionT, RevisionT, BytesT]


class DeviceStatusM:
    t: ClassVar[str] = t(ModelM, RevisionM, RevisionM, BytesM)

    @staticmethod
    def pack(_status: DeviceStatus) -> DeviceStatusT:
        raise NotImplementedError("pack")

    @staticmethod
    def unpack(_status: DeviceStatusT) -> DeviceStatus:
        raise NotImplementedError("unpack")
