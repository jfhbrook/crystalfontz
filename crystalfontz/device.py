from abc import ABC
from collections import defaultdict
from typing import Any, Dict, Self

from crystalfontz.error import DecodeError

DeviceStatus = Any


class Device(ABC):
    LINE_COUNT = 2
    LINE_WIDTH = 16
    ENHANCED_CONTRAST = False

    def status(self: Self, data: bytes) -> DeviceStatus:
        raise NotImplementedError("parse_status")


class CFA533_H1_4_U1_V2(Device):
    LINE_COUNT = 2
    LINE_WIDTH = 16

    # CFA633 accepts one byte between 0 and 200. CFA533 will accept one byte
    # as well, with value between 0 and 50. However, CFA533 will also accept
    # *two* bytes, where the first byte is ignored and the second byte will be
    # used to allow for better resolution.
    ENHANCED_CONTRAST = True

    def status(self: Self, data: bytes) -> DeviceStatus:
        if len(data) != 15:
            raise DecodeError(f"Status expected to be 15 bytes, is {len(data)} bytes")
        # data[0] is reserved
        temp_1 = data[1]
        temp_2 = data[2]
        temp_3 = data[3]
        temp_4 = data[4]
        key_presses = data[5]
        key_releases = data[6]
        atx_power = data[7]
        watchdog_counter = data[8]
        contrast_adjust = data[9]
        backlight = data[10]
        sense_on_floppy = data[11]
        # data[12] is reserved
        cfa633_contrast = data[13]
        backlight = data[14]

        raise NotImplementedError("parse_status")


FirmwareDict = Dict[str, Device]
HardwareDict = Dict[str, FirmwareDict]
DeviceDict = Dict[str, HardwareDict]

DEVICES: DeviceDict = {
    "CFA533": defaultdict(lambda: defaultdict(lambda: CFA533_H1_4_U1_V2()))
}
