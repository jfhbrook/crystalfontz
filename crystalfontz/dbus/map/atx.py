from typing import ClassVar, List, Tuple

from crystalfontz.atx import AtxPowerSwitchFunction, AtxPowerSwitchFunctionalitySettings
from crystalfontz.dbus.map.base import array, t

AtxPowerSwitchFunctionT = str
AtxPowerSwitchFunctionalitySettingsT = Tuple[List[str], bool, float]


class AtxPowerSwitchFunctionM:
    t: ClassVar[str] = "s"

    @staticmethod
    def unpack(function: AtxPowerSwitchFunctionT) -> AtxPowerSwitchFunction:
        return AtxPowerSwitchFunction(function)


class AtxPowerSwitchFunctionalitySettingsM:
    t: ClassVar[str] = t(array(AtxPowerSwitchFunctionM), "bd")

    @staticmethod
    def unpack(
        settings: AtxPowerSwitchFunctionalitySettingsT,
    ) -> AtxPowerSwitchFunctionalitySettings:
        functions, auto_polarity, power_pulse_length = settings
        return AtxPowerSwitchFunctionalitySettings(
            functions={AtxPowerSwitchFunctionM.unpack(name) for name in functions},
            auto_polarity=auto_polarity,
            power_pulse_length_seconds=power_pulse_length,
        )
