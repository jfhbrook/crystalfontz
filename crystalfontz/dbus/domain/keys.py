from typing import ClassVar, Set, Tuple

from crystalfontz.dbus.domain.base import ByteM, OptFloatM, OptFloatT, struct
from crystalfontz.keys import (
    KeyPress,
    KeyState,
    KeyStates,
    KP_DOWN,
    KP_ENTER,
    KP_EXIT,
    KP_LEFT,
    KP_RIGHT,
    KP_UP,
)

KeyPressT = int

KEYPRESSES: Set[KeyPress] = {KP_UP, KP_ENTER, KP_EXIT, KP_LEFT, KP_RIGHT, KP_DOWN}


class KeyPressM(ByteM):
    t: ClassVar[str] = ByteM.t

    @staticmethod
    def unpack(keypress: int) -> KeyPress:
        if keypress not in KEYPRESSES:
            raise ValueError(f"{keypress} is not a valid key press")
        return keypress


KeyStateT = Tuple[bool, bool, bool]
KeyStatesT = Tuple[
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
    KeyStateT,
]


class KeyStateM:
    t: ClassVar[str] = struct("bbb")

    @staticmethod
    def pack(state: KeyState) -> KeyStateT:
        return (state.pressed, state.pressed_since, state.released_since)

    @staticmethod
    def unpack(state: KeyStateT) -> KeyState:
        pressed, pressed_since, released_since = state
        return KeyState(
            pressed=pressed, pressed_since=pressed_since, released_since=released_since
        )


class KeyStatesM:
    t: ClassVar[str] = struct(KeyStateM.t) * 6

    @staticmethod
    def pack(states: KeyStates) -> KeyStatesT:
        return (
            KeyStateM.pack(states.up),
            KeyStateM.pack(states.enter),
            KeyStateM.pack(states.exit),
            KeyStateM.pack(states.left),
            KeyStateM.pack(states.right),
            KeyStateM.pack(states.down),
        )

    @staticmethod
    def unpack(states: KeyStatesT) -> KeyStates:
        up, enter, exit, left, right, down = states
        return KeyStates(
            up=KeyStateM.unpack(up),
            enter=KeyStateM.unpack(enter),
            exit=KeyStateM.unpack(exit),
            left=KeyStateM.unpack(left),
            right=KeyStateM.unpack(right),
            down=KeyStateM.unpack(down),
        )


KeypadBrightnessT = OptFloatT


class KeypadBrightnessM(OptFloatM):
    """
    Map `Optional[float]` to and from `KeypadBrightnessT` (`float`).

    `KeypadBrightnessM` is an alias for `OptFloatM`.
    """

    t: ClassVar[str] = OptFloatM.t
