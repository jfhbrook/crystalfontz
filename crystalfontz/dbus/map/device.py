from typing import ClassVar, Tuple

from crystalfontz.dbus.map.base import struct
from crystalfontz.device import Device


class DeviceM:
    t: ClassVar[str] = struct("sss")

    @staticmethod
    def pack(device: Device) -> Tuple[str, str, str]:
        return (device.model, device.hardware_rev, device.firmware_rev)


class StatusM:
    t: ClassVar[str] = ""
