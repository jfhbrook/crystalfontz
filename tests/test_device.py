import pytest

from crystalfontz.atx import AtxPowerSwitchFunction, AtxPowerSwitchFunctionalitySettings
from crystalfontz.device import CFA533, CFA533Status, Device, DeviceStatus
from crystalfontz.keys import KeyState, KeyStates

KEY_STATES = KeyStates(
    up=KeyState(pressed=False, pressed_since=False, released_since=False),
    enter=KeyState(pressed=True, pressed_since=False, released_since=False),
    exit=KeyState(pressed=False, pressed_since=True, released_since=False),
    left=KeyState(pressed=False, pressed_since=False, released_since=True),
    right=KeyState(pressed=True, pressed_since=True, released_since=False),
    down=KeyState(pressed=True, pressed_since=False, released_since=True),
)


ATX_SETTINGS = AtxPowerSwitchFunctionalitySettings(
    functions={AtxPowerSwitchFunction.KEYPAD_RESET},
    auto_polarity=True,
    reset_invert=False,
    power_invert=False,
    power_pulse_length_seconds=1.0,
)


@pytest.mark.parametrize(
    "device,status,size",
    [
        (
            CFA533(),
            CFA533Status(
                temperature_sensors_enabled={1, 2, 3},
                key_states=KEY_STATES,
                atx_power_switch_functionality_settings=ATX_SETTINGS,
                watchdog_counter=5,
                contrast=0.2,
                keypad_brightness=0.6,
                atx_sense_on_floppy=False,
                cfa633_contrast=0.2,
                lcd_brightness=0.3,
            ),
            15,
        )
    ],
)
def test_status_to_from_bytes(
    device: Device, status: DeviceStatus, size: int, snapshot
) -> None:
    as_bytes = status.to_bytes(device)

    assert len(as_bytes) == size
    assert as_bytes == snapshot

    from_bytes = status.__class__.from_bytes(as_bytes)

    assert from_bytes == status
