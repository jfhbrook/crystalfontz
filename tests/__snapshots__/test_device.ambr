# serializer version: 1
# name: test_status_to_from_bytes[device0-status0]
  b'\x07\x00\x00\x00\x00\x00! \x053<\x00\x00\n\x1e'
# ---
